tests/unit/moduleapi/moduleauth.tcl

```diff
@@ -0,0 +1,405 @@
+
+        # Validate that module auth failed.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        assert_error {*WRONGPASS*} { $rd read }
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+    }
+
+    test {Killing a client in the middle of blocking module auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+        set rd [redis_deferring_client]
+        $rd client id
+        set cid [$rd read]
+
+        # Attempt blocking module auth command on client `cid` and kill the client while module auth
+        # is in progress.
+        $rd AUTH foo pwd
+        wait_for_blocked_clients_count 1
+        r client kill id $cid
+
+        # Validate that the blocked client count goes to 0 and no AUTH command is tracked.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        assert_error {*I/O error reading reply*} { $rd read }
+        assert_match {} [cmdstat auth]
+    }
+
+    test {test RM_AbortBlock Module API during blocking module auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+
+        # Attempt module auth. With the "block_abort" as the password, the "testacl.so" module
+        # blocks the client and uses the RM_AbortBlock API. This should result in module auth
+        # failing and the client being unblocked with the default AUTH err message.
+        assert_error {*WRONGPASS*} {r AUTH foo block_abort}
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+    }
+
+    test {test RM_RegisterAuthCallback Module API during blocking module auth} {
+        r config resetstat
+        r acl setuser foo >defaultpwd on ~* &* +@all
+        set rd [redis_deferring_client]
+
+        # Start the module auth attempt with the standard Redis auth password for the user. This
+        # will result in all module auth cbs attempted and then standard Redis auth will be tried.
+        $rd AUTH foo defaultpwd
+        wait_for_blocked_clients_count 1
+
+        # Validate that we allow modules to register module auth cbs while module auth is already
+        # in progress.
+        assert_equal {OK} [r testmoduleone.rm_register_blocking_auth_cb]
+        assert_equal {OK} [r testmoduletwo.rm_register_auth_cb]
+
+        # Validate that blocking module auth succeeds.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        assert_equal [$rd read] "OK"
+        set stats [cmdstat auth]
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=0} $stats
+
+        # Validate that even the new blocking module auth cb which was registered in the middle of
+        # blocking module auth is attempted - making it take twice the duration (2x 500000 us).
+        regexp "usec_per_call=(\[0-9]{1,})\.*," $stats all usec_per_call
+        assert {$usec_per_call >= 1000000}
+    }
+
+    test {Module unload during blocking module auth} {
```

Comment:
i see a crash in the daily test, maybe someone have the clue to fix it:
https://github.com/enjoy-binbin/redis/actions/runs/5237093743/jobs/9455052489
```
2023-06-11T21:50:08.1436060Z *** [err]: Module unload during blocking module auth in tests/unit/moduleapi/moduleauth.tcl
2023-06-11T21:50:08.1437440Z Expected 'I/O error reading reply' to match '*WRONGPASS*' (context: type source line 89 file /Users/runner/work/redis/redis/tests/support/test.tcl cmd {assert_match $pattern $error $detail} proc ::assert_error level 1) 

2023-06-11T21:49:56.3423960Z ### Starting test Module unload during blocking module auth in tests/unit/moduleapi/moduleauth.tcl
2023-06-11T21:49:56.3424530Z 16727:M 11 Jun 2023 21:49:54.696 * Module 'misc' loaded from /Users/runner/work/redis/redis/tests/modules/misc.so
2023-06-11T21:49:56.3424940Z 16727:M 11 Jun 2023 21:49:54.699 - Accepted 127.0.0.1:55367
2023-06-11T21:49:56.3425310Z 16727:M 11 Jun 2023 21:49:54.715 # Error unloading module testacl: the module has blocked clients. Please wait for them to be unblocked and try again
2023-06-11T21:49:56.3425660Z 16727:M 11 Jun 2023 21:49:54.717 * Module moduleauthtwo unloaded
2023-06-11T21:49:56.3426150Z 16727:M 11 Jun 2023 21:49:54.717 * Module misc unloaded
2023-06-11T21:49:56.3426350Z 
2023-06-11T21:49:56.3426370Z 
2023-06-11T21:49:56.3426520Z === REDIS BUG REPORT START: Cut & paste starting from here ===
2023-06-11T21:49:56.3426820Z 16727:M 11 Jun 2023 21:49:55.738 * Module testacl unloaded
2023-06-11T21:49:56.3427110Z 16727:M 11 Jun 2023 21:49:55.738 # Redis 255.255.255 crashed by signal: 11, si_code: 1
2023-06-11T21:49:56.3427390Z 16727:M 11 Jun 2023 21:49:55.738 # Accessing address: 0x800930778
2023-06-11T21:49:56.3427680Z 16727:M 11 Jun 2023 21:49:55.738 # Crashed running the instruction at: 0x800930778
2023-06-11T21:49:56.3427890Z 
2023-06-11T21:49:56.3428160Z ------ STACK TRACE ------
2023-06-11T21:49:56.3428360Z EIP:
2023-06-11T21:49:56.3428540Z 0x800930778 <_fini+0x800558970> at ???
2023-06-11T21:49:56.3428680Z 
2023-06-11T21:49:56.3428750Z Backtrace:
2023-06-11T21:49:56.3428980Z 0x80047158e <pthread_sigmask+0x54e> at /lib/libthr.so.3
2023-06-11T21:49:56.3429270Z 0x800470b3f <pthread_setschedparam+0x82f> at /lib/libthr.so.3
2023-06-11T21:49:56.3429440Z 
2023-06-11T21:49:56.3429620Z ------ REGISTERS ------
2023-06-11T21:49:56.3429830Z 16727:M 11 Jun 2023 21:49:55.740 # 
2023-06-11T21:49:56.3430040Z RAX:0000000000000000 RBX:0000000802400008
2023-06-11T21:49:56.3430260Z RCX:0000000000000000 RDX:0000000000000000
2023-06-11T21:49:56.3430470Z RDI:00000008008ca288 RSI:0000000000000000
2023-06-11T21:49:56.3430700Z RBP:00007fffd97ccfc0 RSP:00007fffd97ccf90
2023-06-11T21:49:56.3430920Z R8 :0000000000000000 R9 :0000000000000000
2023-06-11T21:49:56.3431110Z R10:0000000000000000 R11:0000000000000246
2023-06-11T21:49:56.3431310Z R12:0000000802401160 R13:0000000000000002
2023-06-11T21:49:56.3431500Z R14:0000000800c4e460 R15:0000000800dce913
2023-06-11T21:49:56.3431700Z RIP:0000000800930778 EFL:0000000000010246
2023-06-11T21:49:56.3431900Z CSGSFS:0000000000000043
2023-06-11T21:49:56.3432110Z ===== End of server log (pid 16727) =====
```